File: 347-top-k-frequent-elements/top-k-frequent-elements.cpp
class Solution {
public:

     
    vector<int> topKFrequent(vector<int>& nums, int k) {




        auto cmp = [](const pair<int,int> & a, const pair<int,int> & b){
            return a.second > b.second;
        };

        unordered_map<int,int> freq;
        for(int num :nums)
        {
            freq[num]++;
        }
        priority_queue<pair<int,int> , vector<pair<int,int>>, decltype(cmp)> pq(cmp);

        for(auto& [num, fr] : freq)
        {
            
            pq.push({num, fr});
            if(pq.size() > k){
                pq.pop();
            }
        }
        // vector<int> ans(pq.begin(),pq.end());
        vector<int> ans;
        while(!pq.empty())
        {
            ans.push_back(pq.top().first);
            pq.pop();
        }
        return ans;
    }
};

File: 347-top-k-frequent-elements/top-k-frequent-elements.java
class Solution {
    public int[] topKFrequent(int[] nums, int k) {
        
        record Pair(int first, int second){};

        HashMap<Integer,Integer> map = new HashMap<>();
        Queue<Pair> pq = new PriorityQueue<>((a,b) -> a.second() - b.second() );

        for(int n : nums)
        {
            map.put(n , map.getOrDefault(n,0) + 1);
        }
        for(Integer key : map.keySet())
        {
            pq.offer(new Pair(key, map.get(key)));

            while(pq.size() > k)
            {
                pq.remove();
            }
        }

        int[] ans = new int[k];
        int i = 0;
        while(!pq.isEmpty())
        {
            ans[i] = pq.remove().first();
            i++;
        }

        return ans;
    }
}